README.rst
*fMRIPrep*: A Robust Preprocessing Pipeline for fMRI Data
=========================================================
*fMRIPrep* is a *NiPreps (NeuroImaging PREProcessing toolS)* application
(`www.nipreps.org <https://www.nipreps.org>`__) for the preprocessing of
task-based and resting-state functional MRI (fMRI).

.. image:: https://img.shields.io/badge/docker-nipreps/fmriprep-brightgreen.svg?logo=docker&style=flat
  :target: https://hub.docker.com/r/nipreps/fmriprep/tags/
  :alt: Docker image available!

.. image:: https://codeocean.com/codeocean-assets/badge/open-in-code-ocean.svg
  :target: https://doi.org/10.24433/CO.ed5ddfef-76a3-4996-b298-e3200f69141b
  :alt: Available in CodeOcean!

.. image:: https://circleci.com/gh/nipreps/fmriprep/tree/master.svg?style=shield
  :target: https://circleci.com/gh/nipreps/fmriprep/tree/master

.. image:: https://readthedocs.org/projects/fmriprep/badge/?version=latest
  :target: http://fmriprep.readthedocs.io/en/latest/?badge=latest
  :alt: Documentation Status

.. image:: https://img.shields.io/pypi/v/fmriprep.svg
  :target: https://pypi.python.org/pypi/fmriprep/
  :alt: Latest Version

.. image:: https://img.shields.io/badge/doi-10.1038%2Fs41592--018--0235--4-blue.svg
  :target: https://doi.org/10.1038/s41592-018-0235-4
  :alt: Published in Nature Methods

.. image:: https://img.shields.io/badge/RRID-SCR__016216-blue.svg
  :target: https://doi.org/10.1038/s41592-018-0235-4
  :alt: RRID:SCR_016216

About
-----

.. image:: https://github.com/oesteban/fmriprep/raw/38a63e9504ab67812b63813c5fe9af882109408e/docs/_static/fmriprep-workflow-all.png


*fMRIPrep* is a functional magnetic resonance imaging (fMRI) data
preprocessing pipeline that is designed to provide an easily accessible,
state-of-the-art interface that is robust to variations in scan acquisition
protocols and that requires minimal user input, while providing easily
interpretable and comprehensive error and output reporting.
It performs basic processing steps (coregistration, normalization, unwarping,
noise component extraction, segmentation, skullstripping etc.) providing
outputs that can be easily submitted to a variety of group level analyses,
including task-based or resting-state fMRI, graph theory measures, surface or
volume-based statistics, etc.

.. note::

   *fMRIPrep* performs minimal preprocessing.
   Here we define 'minimal preprocessing'  as motion correction, field
   unwarping, normalization, bias field correction, and brain extraction.
   See the `workflows section of our documentation
   <https://fmriprep.readthedocs.io/en/latest/workflows.html>`__ for more details.

The *fMRIPrep* pipeline uses a combination of tools from well-known software
packages, including FSL_, ANTs_, FreeSurfer_ and AFNI_.
This pipeline was designed to provide the best software implementation for each
state of preprocessing, and will be updated as newer and better neuroimaging
software become available.

This tool allows you to easily do the following:

- Take fMRI data from raw to fully preprocessed form.
- Implement tools from different software packages.
- Achieve optimal data processing quality by using the best tools available.
- Generate preprocessing quality reports, with which the user can easily
  identify outliers.
- Receive verbose output concerning the stage of preprocessing for each
  subject, including meaningful errors.
- Automate and parallelize processing steps, which provides a significant
  speed-up from typical linear, manual processing.

More information and documentation can be found at
https://fmriprep.readthedocs.io/

Principles
----------
*fMRIPrep* is built around three principles:

1. **Robustness** - The pipeline adapts the preprocessing steps depending on
   the input dataset and should provide results as good as possible
   independently of scanner make, scanning parameters or presence of additional
   correction scans (such as fieldmaps).
2. **Ease of use** - Thanks to dependence on the BIDS standard, manual
   parameter input is reduced to a minimum, allowing the pipeline to run in an
   automatic fashion.
3. **"Glass box"** philosophy - Automation should not mean that one should not
   visually inspect the results or understand the methods.
   Thus, *fMRIPrep* provides visual reports for each subject, detailing the
   accuracy of the most important processing steps.
   This, combined with the documentation, can help researchers to understand
   the process and decide which subjects should be kept for the group level
   analysis.

Citation
--------
**Citation boilerplate**.
Please acknowledge this work using the citation boilerplate that *fMRIPrep* includes
in the visual report generated for every subject processed.
For a more detailed description of the citation boilerplate and its relevance,
please check out the
`NiPreps documentation <https://www.nipreps.org/intro/transparency/#citation-boilerplates>`__.

**Plagiarism disclaimer**.
The boilerplate text is public domain, distributed under the
`CC0 license <https://creativecommons.org/publicdomain/zero/1.0/>`__,
and we recommend *fMRIPrep* users to reproduce it verbatim in their works.
Therefore, if reviewers and/or editors raise concerns because the text is flagged by automated
plagiarism detection, please refer them to the *NiPreps* community and/or the note to this
effect in the `boilerplate documentation page <https://www.nipreps.org/intro/transparency/#citation-boilerplates>`__.

**Papers**.
*fMRIPrep* contributors have published two relevant papers:
`Esteban et al. (2019) <https://doi.org/10.1038/s41592-018-0235-4>`__
[`preprint <https://doi.org/10.1101/306951>`__], and
`Esteban et al. (2020) <https://doi.org/10.1038/s41596-020-0327-3>`__
[`preprint <https://doi.org/10.1101/694364>`__].

**Other**.
Other materials that have been generated over time include the
`OHBM 2018 software demonstration <https://effigies.github.io/fmriprep-demo/>`__
and some conference posters:

* Organization for Human Brain Mapping 2018
  (`Abstract <https://ww5.aievolution.com/hbm1801/index.cfm?do=abs.viewAbs&abs=1321>`__;
  `PDF <https://files.aievolution.com/hbm1801/abstracts/31779/2035_Markiewicz.pdf>`__)

.. image:: _static/OHBM2018-poster_thumb.png
   :target: _static/OHBM2018-poster.png

* Organization for Human Brain Mapping 2017
  (`Abstract <https://ww5.aievolution.com/hbm1701/index.cfm?do=abs.viewAbs&abs=4111>`__;
  `PDF <https://f1000research.com/posters/6-1129>`__)

.. image:: _static/OHBM2017-poster_thumb.png
   :target: _static/OHBM2017-poster.png

Acknowledgements
----------------
This work is steered and maintained by the `NiPreps Community <https://www.nipreps.org>`__.

This work was supported by the Laura and John Arnold Foundation,
the NIH (grant NBIB R01EB020740, PI: Ghosh),
and NIMH (R24MH114705, R24MH117179, R01MH121867, PI: Poldrack)


CHANGES.rst
21.0.0 (December 14, 2021)
==========================
A new series is finally here!

Some highlights of this version's new features include a substantial rework of fMRIPrep's
susceptibility distortion correction (including the addition of FSL's `topup` as the new
default PEPOLAR correction technique), improved flexibility for multiecho scans, and
dependency version upgrades within the Docker environment.

.. note::

  To follow BIDS Derivatives more closely, *fMRIPrep*'s default output layout has changed.
  To enable the legacy layout, use the `--output-layout legacy` flag.
  For more information, please see https://fmriprep.org/en/latest/outputs.html#layout

.. caution::

    As with all minor version increments, working directories
    from previous versions **should not be reused**.

Thank you for using *fMRIPrep*!
If you encounter any issues with this release, please let us know
by posting an issue on our GitHub page!

A full list of changes can be found below.

* DOC: Remove mention to ``epidewarp.fsl`` from ``NOTICE`` (#2629)
* DOC: Update description of output layout, add discussion of legacy mode (#2646)
* DOC: ME | Add missing output to documentation, improve boilerplate (#2608)
* DOC: Add --index-metadata to `pybids layout` faq (#2546)
* DOC: Add warning about slice timing correction in output documentation. (#2502)
* DOC: Update sample report (#2519)
* DOC: Refactor README and citing information (#2474)
* DOC: Point documentation at *SDCFlows* and remove SDC section (#2470)\
* DOC: Transfer duplicated documentation to www.nipreps.org (#2469)
* DOC: Better explanation on how spike regressors are generated (#2465)
* DOC: Clarify that ``res-2`` entity does not mean 2mm (#2466)
* DOC: adds --mem metavar (#2378)
* DOC: skull stripping is forced by default (#2345)
* DOCKER: Shrink Docker image size (#2551)
* DOCKER: Strip ABI tag from libQt5Core.so.5 to prevent Singularity failures (#2535)
* ENH: Make bids the default output layout option (#2555)
* ENH: Mask multi-echo data with reference mask instead of echo-specific masks (#2349)
* ENH: Replace initial FLIRT with mri_coreg, use -basescale 1 for FLIRT-BBR (#2625)
* ENH: Add flag for STC reference time and set in all cases (#2520)
* ENH: Use ``BIDSLayoutIndexer`` and do not index unnecessary modalities (#2494)
* ENH: Allow users to download templateflow templates to directory (#2482)
* ENH: Detect 3D & too-short BOLD series, warn and skip run's workflow building (#2461)
* ENH: Modernize Dockerfile (#2409) (#2270)
* ENH: Leverage SDCFlows 2.x API (#2392)
* ENH: Tedana-friendly ME derivatives (#2575)
* ENH: Better integration of *SDCFlows*' unwarping (#2576)
* ENH: Refactor workflow to adopt *SDCFlows*' new correction workflow (#2547)
* FIX: Threshold boldref resampled values to be non-negative (#2630)
* FIX: Restore SyN-SDC (#2530)
* FIX: Confusing I/O field ``bold_mask`` in bold-to-bold resampling workflow (#2611)
* FIX: ME | Revise SDC compute graph (#2610)
* FIX: Propagate the SDC warp to resampling node also with ME (#2609)
* FIX: Optimize build job on CircleCI (#2605)
* FIX: Minor multiecho issues (#2574)
* FIX: Missing numpy import (#2533)
* FIX: Avoid eager imports from *NiWorkflows* causing API breaks (#2495)
* FIX: Split aCompCor into aCompCor, cCompCor, wCompCor (#2523)
* FIX: Avoid unnecessary connections based on branching logic (#2508)
* FIX: Permit missing TR to show PyBIDS error at workflow construction time (#2513)
* FIX: Simplify STC logic for too short BOLD series (#2489)
* FIX: Catch FreeSurfer error related to FIPS being enabled (#2490)
* FIX: ``DerivativesDataSink`` nondeterministic checksums fixed with ``niworkflows=~1.3.4`` (#2458)
* FIX: Address dependency incompatibilities by pinning specific packages (#2463)
* FIX: Unprotected import of ``sentry_sdk``, which is not a dependency (#2460)
* FIX: Clarify phase encoding direction, rather than axis (#2302)
* MAINT: Conform fast-track outputs to BIDS ordering (#2663)
* MAINT: Reduce redundant processing of ds005 (#2649)
* MAINT: Finalize the work we started with the refactor of the Docker image (#2601)
* MAINT: Run full workflows on tags (#2593)
* MAINT: Stage Python environment of Docker image from nipreps/miniconda (#2581)
* MAINT: Stop NeuroDebian dependency in Dockerfile (#2578)
* MAINT: Run pytest through GitHub actions (#2529)
* MAINT: Relicense +20.3.x - BSD-3-Clause -> Apache License 2.0 (#2325)
* MAINT: Add missing OASIS30 WM/BS probsegs (#2471)
* MAINT: Update BIDS validator to 1.8.0 (#2443)
* MAINT: Address the problems of a sloppy merge (#2468)
* MAINT: Containers - remove ``/root/.npm`` (#2464)
* MAINT: Remove traces of Xenial base (#2459)
* MAINT: Revise CircleCI's cache tags to avoid mixups between builds (#2457)
* MAINT: Back port CircleCI configuration from dev branch (#2456)
* MAINT: Increase minimum tedana version (#2366)
* MAINT: Docker hub has disabled password authentication (#2448)
* MAINT: Add Lea Waller to ``contributors.json`` (#2304)
* MAINT: Move away from nearly deprecated images (#2295)
* MAINT: Migrate from poldracklab to nipreps orgs (#2290)
* MAINT: Ease CI packaging tests (#2472)
* RF/FIX: Iterate over echo indices, not filenames, simplifying iteration logic (#2651)

20.2.5 (October 12, 2021)
=========================
Bug-fix release in the 20.2.x LTS series.

This release includes a fix to `--slice-time-ref` parsing. Also, some
issues in anatomical processing are resolved, including poorly-interpolated
labels in aseg segmentations, and probabilistic segmentations have reverted
to FAST.

  * FIX: --slice-time-ref option parsing (#2573)
  * CI: Add style checks (missing since Travis got throttled) (#2570)

20.2.4 (October 04, 2021)
=========================
Bug-fix release in the 20.2.x LTS series.

This release includes some fixes to minor bugs, and includes enhancements
to allow the user to specify a different slice-timing-correction (STC) target.

Each BOLD series will contain ``SliceTimingCorrected`` metadata to indicate

fmriprep/utils/bids.py
# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
# vi: set ft=python sts=4 ts=4 sw=4 et:
#
# Copyright 2021 The NiPreps Developers <nipreps@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Utilities to handle BIDS inputs."""
import os
import sys
import json
from pathlib import Path


def write_bidsignore(deriv_dir):
    bids_ignore = (
        "*.html", "logs/", "figures/",  # Reports
        "*_xfm.*",  # Unspecified transform files
        "*.surf.gii",  # Unspecified structural outputs
        # Unspecified functional outputs
        "*_boldref.nii.gz", "*_bold.func.gii",
        "*_mixing.tsv", "*_AROMAnoiseICs.csv", "*_timeseries.tsv",
    )
    ignore_file = Path(deriv_dir) / ".bidsignore"

    ignore_file.write_text("\n".join(bids_ignore) + "\n")


def write_derivative_description(bids_dir, deriv_dir):
    from ..__about__ import __version__, DOWNLOAD_URL

    bids_dir = Path(bids_dir)
    deriv_dir = Path(deriv_dir)
    desc = {
        'Name': 'fMRIPrep - fMRI PREProcessing workflow',
        'BIDSVersion': '1.4.0',
        'DatasetType': 'derivative',
        'GeneratedBy': [{
            'Name': 'fMRIPrep',
            'Version': __version__,
            'CodeURL': DOWNLOAD_URL,
        }],
        'HowToAcknowledge':
            'Please cite our paper (https://doi.org/10.1038/s41592-018-0235-4), '
            'and include the generated citation boilerplate within the Methods '
            'section of the text.',
    }

    # Keys that can only be set by environment
    if 'FMRIPREP_DOCKER_TAG' in os.environ:
        desc['GeneratedBy'][0]['Container'] = {
            "Type": "docker",
            "Tag": f"nipreps/fmriprep:{os.environ['FMRIPREP_DOCKER_TAG']}"
        }
    if 'FMRIPREP_SINGULARITY_URL' in os.environ:
        desc['GeneratedBy'][0]['Container'] = {
            "Type": "singularity",
            "URI": os.getenv('FMRIPREP_SINGULARITY_URL')
        }

    # Keys deriving from source dataset
    orig_desc = {}
    fname = bids_dir / 'dataset_description.json'
    if fname.exists():
        orig_desc = json.loads(fname.read_text())

    if 'DatasetDOI' in orig_desc:
        desc['SourceDatasets'] = [{
            'URL': f'https://doi.org/{orig_desc["DatasetDOI"]}',
            'DOI': orig_desc['DatasetDOI']
        }]
    if 'License' in orig_desc:
        desc['License'] = orig_desc['License']

    Path.write_text(deriv_dir / 'dataset_description.json', json.dumps(desc, indent=4))


def validate_input_dir(exec_env, bids_dir, participant_label):
    # Ignore issues and warnings that should not influence FMRIPREP
    import tempfile
    import subprocess
    validator_config_dict = {
        "ignore": [
            "EVENTS_COLUMN_ONSET",
            "EVENTS_COLUMN_DURATION",
            "TSV_EQUAL_ROWS",
            "TSV_EMPTY_CELL",
            "TSV_IMPROPER_NA",
            "VOLUME_COUNT_MISMATCH",
            "BVAL_MULTIPLE_ROWS",
            "BVEC_NUMBER_ROWS",
            "DWI_MISSING_BVAL",
            "INCONSISTENT_SUBJECTS",
            "INCONSISTENT_PARAMETERS",
            "BVEC_ROW_LENGTH",
            "B_FILE",
            "PARTICIPANT_ID_COLUMN",
            "PARTICIPANT_ID_MISMATCH",
            "TASK_NAME_MUST_DEFINE",
            "PHENOTYPE_SUBJECTS_MISSING",
            "STIMULUS_FILE_MISSING",
            "DWI_MISSING_BVEC",
            "EVENTS_TSV_MISSING",
            "TSV_IMPROPER_NA",
            "ACQTIME_FMT",
            "Participants age 89 or higher",
            "DATASET_DESCRIPTION_JSON_MISSING",
            "FILENAME_COLUMN",
            "WRONG_NEW_LINE",
            "MISSING_TSV_COLUMN_CHANNELS",
            "MISSING_TSV_COLUMN_IEEG_CHANNELS",
            "MISSING_TSV_COLUMN_IEEG_ELECTRODES",
            "UNUSED_STIMULUS",
            "CHANNELS_COLUMN_SFREQ",
            "CHANNELS_COLUMN_LOWCUT",
            "CHANNELS_COLUMN_HIGHCUT",
            "CHANNELS_COLUMN_NOTCH",
            "CUSTOM_COLUMN_WITHOUT_DESCRIPTION",
            "ACQTIME_FMT",
            "SUSPICIOUSLY_LONG_EVENT_DESIGN",
            "SUSPICIOUSLY_SHORT_EVENT_DESIGN",
            "MALFORMED_BVEC",
            "MALFORMED_BVAL",
            "MISSING_TSV_COLUMN_EEG_ELECTRODES",
            "MISSING_SESSION"
        ],
        "error": ["NO_T1W"],
        "ignoredFiles": ['/dataset_description.json', '/participants.tsv']
    }
    # Limit validation only to data from requested participants
    if participant_label:
        all_subs = set([s.name[4:] for s in bids_dir.glob('sub-*')])
        selected_subs = set([s[4:] if s.startswith('sub-') else s
                             for s in participant_label])
        bad_labels = selected_subs.difference(all_subs)
        if bad_labels:
            error_msg = 'Data for requested participant(s) label(s) not found. Could ' \
                        'not find data for participant(s): %s. Please verify the requested ' \
                        'participant labels.'
            if exec_env == 'docker':
                error_msg += ' This error can be caused by the input data not being ' \
                             'accessible inside the docker container. Please make sure all ' \
                             'volumes are mounted properly (see https://docs.docker.com/' \
                             'engine/reference/commandline/run/#mount-volume--v---read-only)'
            if exec_env == 'singularity':
                error_msg += ' This error can be caused by the input data not being ' \
                             'accessible inside the singularity container. Please make sure ' \
                             'all paths are mapped properly (see https://www.sylabs.io/' \
                             'guides/3.0/user-guide/bind_paths_and_mounts.html)'
            raise RuntimeError(error_msg % ','.join(bad_labels))

        ignored_subs = all_subs.difference(selected_subs)
        if ignored_subs:
            for sub in ignored_subs:
                validator_config_dict["ignoredFiles"].append("/sub-%s/**" % sub)
    with tempfile.NamedTemporaryFile(mode='w+', suffix='.json') as temp:
        temp.write(json.dumps(validator_config_dict))
        temp.flush()
        try:
            subprocess.check_call(['bids-validator', str(bids_dir), '-c', temp.name])
        except FileNotFoundError:
            print("bids-validator does not appear to be installed", file=sys.stderr)
